htlcswitch/interceptable_switch.go

```diff
@@ -309,66 +309,70 @@ func (s *InterceptableSwitch) ForwardPackets(linkQuit chan struct{}, isReplay bo
 func (s *InterceptableSwitch) interceptForward(packet *htlcPacket,
 	isReplay bool) bool {
 
-	// Process normally if an interceptor is not required and not
-	// registered.
-	if !s.requireInterceptor && s.interceptor == nil {
-		return false
-	}
-
 	switch htlc := packet.htlc.(type) {
 	case *lnwire.UpdateAddHTLC:
 		// We are not interested in intercepting initiated payments.
 		if packet.incomingChanID == hop.Source {
 			return false
 		}
 
-		inKey := channeldb.CircuitKey{
-			ChanID: packet.incomingChanID,
-			HtlcID: packet.incomingHTLCID,
-		}
-
-		// Ignore already held htlcs.
-		if _, ok := s.holdForwards[inKey]; ok {
-			return true
-		}
-
 		intercepted := &interceptedForward{
 			htlc:       htlc,
 			packet:     packet,
 			htlcSwitch: s.htlcSwitch,
 		}
 
-		if s.interceptor == nil && !isReplay {
-			// There is no interceptor registered, we are in
-			// interceptor-required mode, and this is a new packet
-			//
-			// Because the interceptor has never seen this packet
-			// yet, it is still safe to fail back. This limits the
-			// backlog of htlcs when the interceptor is down.
-			err := intercepted.FailWithCode(
-				lnwire.CodeTemporaryChannelFailure,
-			)
-			if err != nil {
-				log.Errorf("Cannot fail packet: %v", err)
-			}
+		return s.forward(intercepted, isReplay)
 
-			return true
-		}
+	default:
+		return false
+	}
+}
 
-		s.holdForwards[inKey] = intercepted
+// forward records the intercepted htlc and forwards it to the interceptor.
+func (s *InterceptableSwitch) forward(
+	fwd contractcourt.InterceptedForward, isReplay bool) bool {
 
-		// If there is no interceptor registered, we must be in
-		// interceptor-required mode. The packet is kept in the queue
-		// until the interceptor registers itself.
-		if s.interceptor != nil {
-			s.sendForward(intercepted)
+	// Process normally if an interceptor is not required and not
+	// registered.
+	if !s.requireInterceptor && s.interceptor == nil {
+		return false
+	}
+
+	inKey := fwd.Packet().IncomingCircuit
+
+	// Ignore already held htlcs.
+	if _, ok := s.holdForwards[inKey]; ok {
+		return true
+	}
+
+	if s.interceptor == nil && !isReplay {
+		// There is no interceptor registered, we are in
+		// interceptor-required mode, and this is a new packet
+		//
+		// Because the interceptor has never seen this packet
+		// yet, it is still safe to fail back. This limits the
+		// backlog of htlcs when the interceptor is down.
+		err := fwd.FailWithCode(
+			lnwire.CodeTemporaryChannelFailure,
+		)
+		if err != nil {
+			log.Errorf("Cannot fail packet: %v", err)
 		}
 
 		return true
+	}
 
-	default:
-		return false
+	s.holdForwards[inKey] = fwd
+
+	// If there is no interceptor registered, we must be in
```

Comment:
Comment doesn't parse? If `interceptor!=nil` then an interceptor is registered?

Reply:
Fixed